Add Mautic 3 upgrade script

dennisameling opened the pull request:
Provides a new upgrade script to allow for upgrading M2 to M3. See https://github.com/mautic/mautic/issues/8819 for details

![image](https://user-images.githubusercontent.com/17739158/99320978-12fc4100-286d-11eb-8791-5a55ce876232.png)

**Steps to test**
Mautic 2.16.3 will add a separate upgrade script for the upgrade to Mautic 3. You will first install the Mautic 2.16.3 beta and then upgrade to Mautic 3.

1. You have three options to test this PR:
**Clone this PR**: clone this PR and go to step 3 (assuming that you have a working 2.x installation already. If you clone in an empty folder, then also follow step 2).
**Clean installation**: Download the custom [2.16.3-beta.zip package](https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta.zip)
**Upgrade existing 2.x installation**: Download the custom [2.16.3-beta-update.zip package](https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta-update.zip)

2. Install 2.16.3-beta (clean install or run `app/console mautic:update:apply --update-package=2.16.3-beta-update.zip` to upgrade an existing installation)

3. You have two options for testing the upgrade to M3 (please do both if you have time):
**UI upgrade**: go to your Mautic dashboard, you should see a reguler update notification like you're used to. If not, run `app/console mautic:update:find` to force search for new updates.
**CLI upgrade**: run `app/console mautic:update:find` and if the update is found, run `php upgrade_V3.php`
4. You should now be able to start the upgrade. Please share your experience with the upgrade script here!

**Testing with DDEV**
For everyone who's using DDEV, you can take these steps to test:

- Open a terminal
- `mkdir mautic3upgrade`
- `cd mautic3upgrade`
- Run `wget https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta.zip`
- Run `unzip 2.16.3-beta.zip`
- Run `ddev config --webserver-type apache-fpm --php-version 7.3`
- Run `ddev start`, and you should be good to go :rocket:

dennisameling committed:
First work on upgrade script

dennisameling committed:
Almost finish work on browser-based upgrade

dennisameling committed:
Temp change in Mautic 3 download URL

Had to create a new package due to the issue mentioned here: https://github.com/mautic/mautic/issues/8826

luizeof commented:
Hello @dennisameling, how are you? Is this work still in progress? I think of bundle it on the docker-mautic

dennisameling commented:
@luizeof yes, hope to have this finished by tomorrow, but partly blocked by https://github.com/mautic/mautic/issues/8831 as the migration of all relevant settings needs to be part of the script as well. Will update this PR/issue as soon as I have more!

luizeof commented:
could you include this #8834 in your script please?

dennisameling commented:
I'll need to check if that's something the upgrade script could fix or whether it's a Mautic 3 bug. In the meantime, can you try setting up a fresh M3 instance, then try to set the trusted proxy settings in the UI rather than directly adding it to local.php? It could be that the way it's written in the local.php is slightly different in M3 compared to M2. Feel free to report back the results in #8834 and @ me over there to follow up.

dennisameling committed:
Build and test CLI version of upgrade, build file cleanup after upgrade

dennisameling committed:
Replace config default values with M3 default values, see #8831

dennisameling committed:
Finish pre-upgrade checks in M3 upgrade script

dennisameling committed:
Add mysqldump database backup + add pre-upgrade warning about api_rate_limiter_cache + fix bugs

dennisameling committed:
Add check for custom plugins, move restoring user data to end of upgrade

dennisameling committed:
Add upgrade log with extensive logging, add error codes, capture Symfony commands output, etc.

dennisameling committed:
Remove unnecessary code

dennisameling committed:
Refactor UI upgrade using JSON/jQuery, several fixes

dennisameling committed:
Run DB migrations one by one in UI, remove GIF

dennisameling committed:
Add tests to production package to prevent CLI warnings

dennisameling committed:
Build integration with Mautic native update functionality in UI and CLI

luizeof commented:
@dennisameling will this completely replace the current upgrade.php and also the updater located on the mautic interface bell?

dennisameling committed:
Cache warmup, switch to JSON upgrade config, phpcs

dennisameling committed:
Send error codes in telemetry, fix UI upgrade bug

dennisameling commented:
@luizeof It's a completely separate (standalone) upgrade script (`upgrade_v3.php`), because of the extensive file system changes in M3. It also has an integration with Mautic's native update functionality. So, the user will get a notification in the UI (notification bell) that the upgrade to M3 is available, and when they click "Start update" they are redirected to `upgrade_v3.php`. When running `app/console mautic:update:find` (CLI) you will be instructed on how to start the `upgrade_v3.php` script. If you want to see what that looks like, see https://github.com/mautic/mautic/issues/8819#issuecomment-640244012 :)

dennisameling committed:
Update docs URL for error codes

dennisameling commented:
One thing I just found while testing:

On systems with `open_basedir` active, the following error is shown:

![image](https://user-images.githubusercontent.com/17739158/83980561-1cc16600-a917-11ea-8665-2c6bfee2153b.png)

```
[2020-06-07 21:31:13] ERR_BUILD_M3_CACHE: Failed to build cache for Mautic 3. All your data has been successfully migrated to Mautic 3, but this error very likely needs to be fixed before you can start using Mautic 3. <br><br>Command output: 
 // Clearing the cache for the prod environment with debug                      
 // true                                                                        


In Kernel.php line 363:
                                                                               
  Warning: file_exists(): open_basedir restriction in effect. File(/var/www/v  
  hosts/composer.json) is not within the allowed path(s): (/var/www/vhosts/mysit 
  e.nl/:/tmp/)                                                  
                                                                               

cache:clear [--no-warmup] [--no-optional-warmers] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command>
```

This is probably due to the fact that `composer.json` is not included in the M3 package, but deleted in the upgrade package script. Only the `vendor` folder is left.

The script that is executed in this step is `php bin/console cache:clear --no-interaction --env=prod --no-debug`.

dennisameling committed:
Prevent undefined variable in NotificationModel

If the user has disableUpdates turned on, $isMautic3Upgrade is undefined. This commit fixes that by moving initial setup of the variable out of the conditional logic.

kuzmany commented:
Upgrade scripts looks like great of piece software with a lot of hardworking time.

My upgrade test experiences:

I did test on clean installation of 2.16.3

### windows localhost enviroment

#### web

- is_mysqldump_available() return true but backup cannot provide, because windows don't know mysqlbackup
- after comment that part I was able to upgrade successfully  👍 
- just clear cache hit timeout - I increased $minExecutionTime to 900

#### console 

I was not able to do upgrade

### linux enviroment produtions (serverpilot)

#### web

Upgrade successful 👍 


#### console 
I was not able to do upgrade

```
Getting the amount of available Mautic 2 database migrations...
string(59) "Couldn't determine amount of available database migrations."
ERR_MAUTIC_2_MIGRATIONS_FAILED: We couldn't reliably detect the amount of available database migrations.
ERR_MAUTIC_2_MIGRATIONS_FAILED: We couldn't reliably detect the amount of available database migrations.. For more details about this message, see https://docs.mautic.org/en/mautic-3-upgrade/mautic-3-upgrade-errors#err_mautic_2_migrations_failed
```

RCheesley commented on upgrade_v3.php:
```diff
@@ -0,0 +1,1974 @@
+
+            if ($success === false) {
+                throwErrorAndWriteToLog(
+                    "ERR_MAUTIC_2_MIGRATIONS_FAILED",
```
I think we should differentiate between being unable to find the migrations and there being a failure with applying them. At the moment we have the same error code for both situations of ERR_MAUTIC_2_MIGRATIONS_FAILED.

```suggestion
                    "ERR_MAUTIC_2_MIGRATIONS_IDENTIFICATION_FAILED",
```

RCheesley replied:
I think they should also apply to the M3 migrations

dennisameling replied:
Done in https://github.com/mautic/mautic/pull/8824/commits/2751b2787e2bad6c463ef59c056aae2ff281b251

RCheesley commented on upgrade_v3.php:
```diff
@@ -0,0 +1,1974 @@
+
+                        if ($success === false) {
+                            throwErrorAndWriteToLog(
+                                "ERR_MAUTIC_2_MIGRATIONS_FAILED",
```
I think this error code should be specific that the application of the migration has failed.

```suggestion
                                "ERR_MAUTIC_2_MIGRATIONS_APPLICATION_FAILED",
```

dennisameling replied:
Not necessary due to the introduction of the `_MIGRATIONS_IDENTIFICATION_FAILED` error codes in https://github.com/mautic/mautic/pull/8824/commits/2751b2787e2bad6c463ef59c056aae2ff281b251

dennisameling committed:
Fix mysqldump checks + add Windows compatibility

dennisameling commented:
Thanks for testing @kuzmany! `mysqldump` can actually work on Windows, as long as it's in your PATH. For me that was `C:\wamp64\bin\mariadb\mariadb10.4.10\bin`. I also had a major bug in my logic, `is_mysqldump_available()` was always returning `true`. Just fixed that in https://github.com/mautic/mautic/pull/8824/commits/c2e0c6d63311588517dfbbe584417853746a2137 😄 will check the other problems tomorrow 👍

dennisameling committed:
Fix database migrations check on CLI upgrades

dennisameling commented:
Just fixed the "Getting the amount of available Mautic 2 database migrations..." on CLI upgrades. Also added a check that if the platform is Windows, set the `max_execution_time` to 600 instead of 240. Building cache really is painfully slow on this platform, but with a `max_execution_time` that's high enough, it succeeds without errors. https://github.com/mautic/mautic/pull/8824/commits/ab5dbb2a503ad9731e118a641fcb427abf08273b

Will work on the other problems later today and then release a new `2.16.3-beta.zip` package for testing 👍

dennisameling committed:
Introduce error codes for identifying db migrat.

dennisameling committed:
add memory_limit check, use mau.tc URL shortener

dennisameling committed:
Move folder movements in CLI to phase 2

dennisameling committed:
Make script more robust on Windows + CLI

dennisameling committed:
Better memory_limit logic, move backupfiles to var

dennisameling committed:
Add remove backup files option at end of upgrade

dennisameling committed:
Add note to check if M3 works before remov. backup

dennisameling commented:
Did a bunch of bug fixes and enhancements yesterday and today (https://github.com/dennisameling/mautic/compare/84983e8...84fcd8a):
- Improve upgrade experience on Windows (higher `max_execution_time`, mysqldump improvements)
- Add warning on Windows that cache creation takes long and users should be patient
- Add memory limit check (needs to be at least 256M, on Windows at least 512M)
- Use mau.tc URL shortener for redirects to the docs
- Add separate error codes `ERR_MAUTIC_2_MIGRATIONS_IDENTIFICATION_FAILED` and `ERR_MAUTIC_3_MIGRATIONS_IDENTIFICATION_FAILED`, these indicate that identifying potential migrations failed
- Display PHP errors on CLI upgrade, to make it easier for folks to troubleshoot
- Move the Mautic 2 to 3 file movements to phase 2 of the CLI upgrade (got errors on Windows prior to fixing that)
- Fix mysqldump for users that don't have a password set on their database
- Add option to remove backup files at the end of the upgrade, which looks like this in the UI:

![image](https://user-images.githubusercontent.com/17739158/84578114-96979a80-adc2-11ea-8da2-eef61d2d7c5e.png)

Also updated [2.16.3-beta.zip](https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta.zip) and [2.16.3-beta-update.zip](https://github.com/dennisameling/mautic/releases/download/2.16.3-beta/2.16.3-beta-update.zip) on their original URLs with the above fixes, so the script is ready to test again 🎉 

cc @kuzmany @RCheesley

RCheesley commented:
Awesome, will do some tests tomorrow! Fantastic work on this!

adiux reviewed (approved):
tested with ddev, from clean 2.16.3-beta
